src/lib/polygon.js

```diff
@@ -0,0 +1,240 @@
+        }
+
+        var imax = pts.length,
+            x1 = pts[0][0],
+            y1 = pts[0][1],
+            crossings = 0,
+            i,
+            x0,
+            y0,
+            xmini,
+            ycross;
+
+        for(i = 1; i < imax; i++) {
+            // find all crossings of a vertical line upward from pt with
+            // polygon segments
+            // crossings exactly at xmax don't count, unless the point is
+            // exactly on the segment, then it counts as inside.
+            x0 = x1;
+            y0 = y1;
+            x1 = pts[i][0];
+            y1 = pts[i][1];
+            xmini = Math.min(x0, x1);
+
+            // outside the bounding box of this segment, it's only a crossing
+            // if it's below the box.
+            if(x < xmini || x > Math.max(x0, x1) || y > Math.max(y0, y1)) {
+                continue;
+            }
+            else if(y < Math.min(y0, y1)) {
+                // don't count the left-most point of the segment as a crossing
+                // because we don't want to double-count adjacent crossings
+                // UNLESS the polygon turns past vertical at exactly this x
+                // Note that this is repeated below, but we can't factor it out
+                // because
+                if(x !== xmini) crossings++;
+            }
+            // inside the bounding box, check the actual line intercept
+            else {
+                // vertical segment - we know already that the point is exactly
+                // on the segment, so mark the crossing as exactly at the point.
+                if(x1 === x0) ycross = y;
+                // any other angle
+                else ycross = y0 + (x - x0) * (y1 - y0) / (x1 - x0);
+
+                // exactly on the edge: counts as inside the polygon, unless it's the
+                // first edge and we're omitting it.
+                if(y === ycross) {
+                    if(i === 1 && omitFirstEdge) return false;
+                    return true;
+                }
+
+                if(y <= ycross && x !== xmini) crossings++;
+            }
+        }
+
+        // if we've gotten this far, odd crossings means inside, even is outside
+        return crossings % 2 === 1;
+    }
+
+    return {
+        xmin: xmin,
+        xmax: xmax,
+        ymin: ymin,
+        ymax: ymax,
+        pts: pts,
+        contains: isRect ? rectContains : contains,
+        isRect: isRect
+    };
+};
+
+/**
+ * Test if a segment of a points array is bent or straight
+ *
+ * @param pts Array of [x, y] pairs
+ * @param start the index of the proposed start of the straight section
+ * @param end the index of the proposed end point
+ * @param tolerance the max distance off the line connecting start and end
+ *      before the line counts as bent
+ * @returns boolean: true means this segment is bent, false means straight
+ */
+var isBent = polygon.isSegmentBent = function isBent(pts, start, end, tolerance) {
```

Comment:
Used to simplify the lasso polygon. This is pretty similar to the scatter line decimation algorithm, so not super :palm_tree: but line decimation has a few other quirks that made it seem annoying to combine.